### PR TITLE
studio: Add Studio authentication to DVC

### PR DESCRIPTION
**Pull request by amritghimire**
* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Adds a new authentication command to DVC that authorizes dvc with Studio. 

Commands added are:
- `dvc studio login`
- `dvc studio logout`
- `dvc studio token`

## Studio command
```sh
usage: dvc studio [-h] [-q | -v] {login,logout,token} ...

Authorize dvc with Studio and set the token. When this is
set, DVC uses this to share live experiments and notify
Studio about pushed experiments.
Documentation: <https://man.dvc.org/studio>

positional arguments:
  {login,logout,token}  Use `dvc studio CMD --help` to display command-specific help.
    login               Authenticate DVC with Studio host
    logout              Logout user from Studio
    token               View the token dvc uses to contact Studio

options:
  -h, --help            show this help message and exit
  -q, --quiet           Be quiet.
  -v, --verbose         Be verbose.
```

## Login

To login, you can call `dvc studio login` which will open the Studio with a device code prefilled in your browser. Once the user authorizes the code, the token is saved to global config.
```sh
usage: dvc studio login [-h] [-q | -v] [-H HOSTNAME] [-s SCOPES] [-n NAME] [-d]

By default, this command authorize dvc with Studio with
 default scopes and a random  name as token name.
Documentation: <https://man.dvc.org/studio/login>

options:
  -h, --help            show this help message and exit
  -q, --quiet           Be quiet.
  -v, --verbose         Be verbose.
  -H HOSTNAME, --hostname HOSTNAME
                        The hostname of the Studio instance to authenticate with.
  -s SCOPES, --scopes SCOPES
                        The scopes for the authentication token.
  -n NAME, --name NAME  The name of the authentication token. It will be used to identify token shown in Studio profile.
  -d, --use-device-code
                        Use authentication flow based on user code. You will be presented with user code to enter in browser. DVC will also use this if it
                        cannot launch browser on your behalf.
```

#### Normal flow
On running the command `dvc studio login` , following screen is presented in CLI:
```sh
dvc studio login
A web browser has been opened at
https://studio.iterative.ai/auth/device-login.
Please continue the login in the web browser.
If no web browser is available or if the web browser fails to open,
use device code flow with `dvc studio login --use-device-code`.
```

A webbrowser is opened with the device code in Studio.
image
Once user authorizes the token in the screen above, the token is saved to the user's global config.
Additional message as below will be shown in the CLI and will exit.
```sh
Authentication successful. The token will be available as risen-geum in Studio profile.
```

#### Device login flow
In case, you are running this in remote machine or where you are unable to open web browser, the following screen is presented.
```
Please open the following url in your browser.
https://studio.iterative.ai/auth/device-login
And enter the user code below 06D21JBK to authorize.
```
Every other flow is same.

## Logout
```sh
usage: dvc studio logout [-h] [-q | -v]

This command helps to log out user from DVC Studio.

Documentation: <https://man.dvc.org/studio/logout>

options:
  -h, --help     show this help message and exit
  -q, --quiet    Be quiet.
  -v, --verbose  Be verbose.
```

## Token
```sh
usage: dvc studio token [-h] [-q | -v]

View the token dvc uses to contact Studio
Documentation: <https://man.dvc.org/studio/token>

options:
  -h, --help     show this help message and exit
  -q, --quiet    Be quiet.
  -v, --verbose  Be verbose.
```

TODO: 
- Needs https://github.com/iterative/studio/pull/8147 to be merged and deployed. 
- `dvc studio status` to check the status and scopes of existing token. This requires changes in Studio first.

Fixes https://github.com/iterative/dvc/issues/9265

### CODE REVIEW

**Comment by dberenbaum** [comment]
Minor, but can we name it `dvc studio` since we already use `studio` as the config section in `.dvc/config`?

When I try it, it always asks me for a device code.

Should `dvc auth logout` try to delete the token in Studio so you don't end up with a lot of stale tokens?

**Comment by amritghimire** [comment]
> Minor, but can we name it `dvc studio` since we already use `studio` as the config section in `.dvc/config`?

Sure. Will update it .

> When I try it, it always asks me for a device code.

It is pending in https://github.com/iterative/studio/pull/8147 . For now, you can copy the device code from the url params as well. 

> Should `dvc auth logout` try to delete the token in Studio so you don't end up with a lot of stale tokens?

User has that option from UI as well. Should we show the message instead?

**Comment by dberenbaum** [comment]
> > Should `dvc auth logout` try to delete the token in Studio so you don't end up with a lot of stale tokens?
> 
> User has that option from UI as well. Should we show the message instead?

What message? Not sure I follow what you mean.

**Comment by amritghimire** [comment]
> > > Should `dvc auth logout` try to delete the token in Studio so you don't end up with a lot of stale tokens?
> > 
> > 
> > User has that option from UI as well. Should we show the message instead?
> 
> What message? Not sure I follow what you mean.

I meant, showing the link to remove the token from UI or show some help message saying you can remove it as well. I think if user used dvc cli to authenticate and used the token else where, do we want to invalidate that as well?

**Comment by dberenbaum** [comment]
> I meant, showing the link to remove the token from UI or show some help message saying you can remove it as well. I think if user used dvc cli to authenticate and used the token else where, do we want to invalidate that as well?

Makes sense, thanks!

**Comment by pmrowla** [comment]
Seems like the internals for this may belong in https://github.com/iterative/dvc-studio-client instead of `utils/studio`

**Comment by amritghimire** [comment]
> Seems like the internals for this may belong in [iterative/dvc-studio-client](https://github.com/iterative/dvc-studio-client) instead of `utils/studio`

Moved those utils to dvc-studio-client. PTAL

**Comment by efiop** [comment]
> Seems like the internals for this may belong in https://github.com/iterative/dvc-studio-client instead of `utils/studio`

@pmrowla Could you take a look at how it is now, please?

**Comment by efiop** [comment]
@skshetry Could you please review this?

File: dvc/commands/studio.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
+import argparse
+import logging
+import os
+
+from funcy import get_in
+
+from dvc.cli.command import CmdBase
+from dvc.cli.utils import append_doc_link, fix_subparsers
+
+logger = logging.getLogger(__name__)
```
Comment (skshetry) [inline dvc/commands/studio.py:10]:
```suggestion
from dvc.log import logger

logger = logger.getChild(__name__)
```

File: dvc/commands/studio.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        except ValueError as e:
+            ui.error_write(str(e))
+            return 1
+
+        access_token = check_token_authorization(uri=token_uri, device_code=device_code)
+        if not access_token:
+            ui.write(
+                "failed to authenticate: This 'device_code' has expired.(expired_token)"
+            )
+            return 1
```
Comment (skshetry) [inline dvc/commands/studio.py:39]: Can we move all of these to `dvc-studio-client` as well (make it a single API) and catch proper exception to notify users? I am not sure we need to deal with these details of opening webbrowsers, etc here. WDYT?

File: dvc/commands/studio.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+        "--name",
+        action="store",
+        default=None,
+        help="The name of the authentication token. It will be used to\n"
+        "identify token shown in Studio profile.",
+    )
+
+    login_parser.add_argument(
+        "-d",
+        "--use-device-code",
```
Comment (skshetry) [inline dvc/commands/studio.py:181]: What does this flag mean? Aren't we using device code like thing even if we don't open a link ourself?

**Review by skshetry — COMMENTED** [review]
I took a brief look here. The code overall looks good, but I don't want to deal with a lot of details here and should be best handled in `dvc-studio-client`. 

Ideally, there should be a single call to `dvc-studio-client` that returns the token, and/or error message with proper message. 
Again, code looks good, but we need to figure out how to organize this and where. :)

File: dvc/commands/studio.py (hunk 3 above)
Comment (amritghimire) [reply dvc/commands/studio.py:181]: It is something similar to what is implemented at https://learn.microsoft.com/en-us/cli/azure/authenticate-azure-cli-interactively

Comment (skshetry) [reply dvc/commands/studio.py:181]: Looks like `azure login` has two different login mechanism, default being Auth Code Flow, and other being Device Code Flow. `--use-device-code` is being used to change that default. Since we only have one type of flow, it does not seem like we need this?

File: tests/unit/utils/test_studio.py [hunk 4]
```diff
@@ -13,7 +13,6 @@
 
 CONFIG = {"offline": True, "repo_url": "repo_url", "token": "token", "url": "url"}
 
-
```
Comment (efiop) [inline tests/unit/utils/test_studio.py:16]: Looks like an unrelated stray change

File: dvc/utils/studio.py [hunk 5]
```diff
@@ -66,7 +66,7 @@ def notify_refs(
     try:
         r = post("webhook/dvc", token, data, base_url=base_url)
     except requests.RequestException as e:
-        logger.trace("", exc_info=True)
+        logger.trace("", exc_info=True)  # type: ignore[attr-defined]
```
Comment (skshetry) [inline dvc/utils/studio.py:69]: The suppression is probably not needed now.

File: tests/func/test_studio.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
+from requests import Response
+
+from dvc.cli import main
+from dvc.commands.studio import DEFAULT_SCOPES
+from dvc.utils.studio import STUDIO_URL
+from tests.unit.command.test_studio import MOCK_RESPONSE
+
+
+def test_auth_expired(mocker, M):
```
Comment (skshetry) [inline tests/func/test_studio.py:9]: @amritghimire, also, please move the most of these tests to `dvc-studio-client`.
And here let's mock the studio-client's API and test for saving config.

File: dvc/commands/studio.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+        from dvc.env import DVC_STUDIO_URL
+        from dvc.ui import ui
+        from dvc.utils.studio import STUDIO_URL
+
+        name = self.args.name
+        hostname = self.args.hostname or os.environ.get(DVC_STUDIO_URL) or STUDIO_URL
+        scopes = self.args.scopes or DEFAULT_SCOPES
+
+        try:
+            token_name, access_token = initiate_authorization(
```
Comment (skshetry) [inline dvc/commands/studio.py:28]: On studio-client side, let's rename this to a more meaningful name. `get_access_token()` perhaps? Or if you have something better.

File: dvc/commands/studio.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+                scopes=scopes,
+                use_device_code=self.args.use_device_code,
+            )
+        except StudioAuthError as e:
+            ui.error_write(str(e))
+            return 1
+
+        self.save_config(hostname, access_token)
+        ui.write(
+            f"Authentication successful. The token will be "
```
Comment (skshetry) [inline dvc/commands/studio.py:40]:
```suggestion
            "Authentication successful. The token will be "
```

File: tests/unit/command/test_studio.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
+from unittest import mock
+
+from dvc_studio_client.auth import AuthorizationExpired
+
+from dvc.cli import main
+from dvc.utils.studio import STUDIO_URL
+
+
+def test_studio_login_invalid_scope():
+    assert main(["studio", "login", "--scopes", "invalid!"]) == 1
```
Comment (skshetry) [inline tests/unit/command/test_studio.py:10]: It looks like dvc is not responsible for checking this scope. Please move this to dvc-studio-client and remove it from here.

File: tests/unit/command/test_studio.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+
+@mock.patch("dvc_studio_client.auth.get_access_token")
+def test_studio_login_token_check_failed(mock_get_access_token):
```
Comment (skshetry) [inline tests/unit/command/test_studio.py:14]:
```suggestion
def test_studio_login_token_check_failed(mocker):
	mock = mocker.patch("dvc_studio_client.auth.get_access_token")
```

We use `mocker` fixture instead of using `@mock.patch`. Also it's better to use simpler names like `mock` instead when there is only a single mock.

File: tests/unit/command/test_studio.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+        hostname="https://example.com",
+        scopes="experiments",
+        use_device_code=False,
+        client_name="dvc",
+    )
+
+
+@mock.patch("dvc_studio_client.auth.get_access_token")
+@mock.patch("webbrowser.open")
+def test_studio_device_code(mock_webbrowser_open, mock_get_access_token):
```
Comment (skshetry) [inline tests/unit/command/test_studio.py:61]: I guess this is not necessary?

File: dvc/commands/studio.py [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+            token_name, access_token = get_access_token(
+                token_name=name,
+                hostname=hostname,
+                scopes=scopes,
+                use_device_code=self.args.use_device_code,
+                client_name="dvc",
+            )
+        except StudioAuthError as e:
+            ui.error_write(str(e))
+            return 1
```
Comment (skshetry) [inline dvc/commands/studio.py:35]: Note to self: I need to look into `dvc-studio-client` if we can improve upon the API, especially around the side effects of opening a browser and printing things.

Comment (skshetry) [reply dvc/commands/studio.py:35]: Also I am having a second thought if webbrowser.open() belongs in dvc instead. :(

**Comment by skshetry** [comment]
@shcheklein, @dberenbaum, could you PTAL as well?

**Review by dberenbaum — APPROVED** [review]

File: dvc/commands/studio.py [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+        name = self.args.name
+        hostname = self.args.hostname or os.environ.get(DVC_STUDIO_URL) or STUDIO_URL
+        scopes = self.args.scopes
+
+        try:
+            token_name, access_token = get_access_token(
+                token_name=name,
+                hostname=hostname,
+                scopes=scopes,
+                use_device_code=self.args.use_device_code,
```
[the remark this answers is not in the document]
Comment (shcheklein) [reply dvc/commands/studio.py:30]: The one that I remember is `az login`, and I think it was not waiting for any input. We can check a few other tools. I don't have a strong opinion on this (means probably I would err to keep it simpler and faster unless we have some security concern, etc)

File: dvc/commands/studio.py [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+        if not token:
+            ui.error_write("Not logged in to Studio.")
+            return 1
+
+        ui.write(token)
+        return 0
+
+
+def add_parser(subparsers, parent_parser):
+    STUDIO_HELP = "Authenticate dvc with Iterative Studio"
```
Comment (shcheklein) [inline dvc/commands/studio.py:80]: dvc -> DVC (and let's spend time to review messages, style, etc)

File: tests/unit/command/test_studio.py [hunk 15]
```diff
@@ -0,0 +1,81 @@
+from dvc_studio_client.auth import AuthorizationExpired
```
Comment (shcheklein) [inline tests/unit/command/test_studio.py:1]: Authorization or Authothentication?

**Review by shcheklein — APPROVED** [review]
Assuming DVC has reviewed and approved it. Thanks @amritghimire

File: dvc/commands/studio.py (hunk 3 above)
Comment (skshetry) [reply dvc/commands/studio.py:142]: Can we change this to `--no-open`/`--no-browser`/`--no-web`, etc?

File: dvc/utils/studio.py [hunk 16]
```diff
@@ -35,7 +35,7 @@ def post(
 
     logger.trace("Sending %s to %s", data, url)
 
-    headers = {"Authorization": f"token {token}"}
+    headers = {"Authorization": f"token {token}"} if token else None
```
Comment (skshetry) [inline dvc/utils/studio.py:38]: Is this still relevant?

File: dvc/commands/studio.py (hunk 3 above)
Comment (amritghimire) [reply dvc/commands/studio.py:142]: I think `--use-device-code` should be fine since this implies user wants to use the device code itself and it matches the behavior of existing tools like `az login`

Comment (skshetry) [reply dvc/commands/studio.py:142]: User has to use device code, no matter what. Again, `az` does not use device code by default and the option is to force that.

Comment (amritghimire) [reply dvc/commands/studio.py]: Updated.

File: dvc/commands/studio.py [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+        "-n",
+        "--name",
+        action="store",
+        default=None,
+        help="The name of the authentication token. It will be used to\n"
+        "identify token shown in Studio profile.",
+    )
+
+    login_parser.add_argument(
+        "-o",
```
Comment (skshetry) [inline dvc/commands/studio.py:141]:
```suggestion
```

File: dvc/commands/studio.py [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+        name = self.args.name
+        hostname = self.args.hostname or os.environ.get(DVC_STUDIO_URL) or STUDIO_URL
+        scopes = self.args.scopes
+
+        try:
+            token_name, access_token = get_access_token(
+                token_name=name,
+                hostname=hostname,
+                scopes=scopes,
+                use_device_code=self.args.no_open,
```
Comment (skshetry) [inline dvc/commands/studio.py:30]: Consider doing the same upstream.

**Review by skshetry — APPROVED** [review]
